File: PythonCodes/Fig05_02.py
#%%
"""
Created on Thu Jan 11 2019
Paths for the Jump diffusion proces of Merton
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GeneratePathsMerton(NoOfPaths,NoOfSteps,S0, T,xiP,muJ,sigmaJ,r,sigma):    

    # Create empty matrices for Poisson and compensated Poisson process

    X = np.zeros([NoOfPaths, NoOfSteps+1])
    S = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
                
    dt = T / float(NoOfSteps)
    X[:,0] = np.log(S0)
    S[:,0] = S0
    
    # Expectation E(e^J) for J~N(muJ,sigmaJ^2)

    EeJ = np.exp(muJ + 0.5*sigmaJ*sigmaJ)
    ZPois = np.random.poisson(xiP*dt,[NoOfPaths,NoOfSteps])
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    J = np.random.normal(muJ,sigmaJ,[NoOfPaths,NoOfSteps])
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])

        # Making sure that samples from a normal have mean 0 and variance 1

        X[:,i+1]  = X[:,i] + (r - xiP*(EeJ-1) - 0.5*sigma*sigma)*dt +sigma*np.sqrt(dt)* Z[:,i]\
                    + J[:,i] * ZPois[:,i]
        time[i+1] = time[i] +dt
        
    S = np.exp(X)
    paths = {"time":time,"X":X,"S":S}
    return paths

def GeneratePathsPoisson(NoOfPaths,NoOfSteps,T,xiP):    

    # Create empty matrices for Poisson and compensated Poisson process

    X = np.zeros([NoOfPaths, NoOfSteps+1])
    Xc = np.zeros([NoOfPaths, NoOfSteps+1])
    time = np.zeros([NoOfSteps+1])
                
    dt = T / float(NoOfSteps)
    
    Z = np.random.poisson(xiP*dt,[NoOfPaths,NoOfSteps])
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        X[:,i+1]  = X[:,i] + Z[:,i]
        Xc[:,i+1] = Xc[:,i] -xiP*dt + Z[:,i]
        time[i+1] = time[i] +dt
        
    paths = {"time":time,"X":X,"Xcomp":Xc}
    return paths

def mainCalculation():
    NoOfPaths = 25
    NoOfSteps = 500
    T = 5
    xiP = 1
    muJ = 0
    sigmaJ = 0.5
    sigma = 0.2
    S0 =100
    r=0.05
    Paths = GeneratePathsMerton(NoOfPaths,NoOfSteps,S0, T,xiP,muJ,sigmaJ,r,sigma)
    timeGrid = Paths["time"]
    X = Paths["X"]
    S = Paths["S"]
           
    plt.figure(1)
    plt.plot(timeGrid, np.transpose(X))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("X(t)")
    
    plt.figure(2)
    plt.plot(timeGrid, np.transpose(S))   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("S(t)")
    
                       
mainCalculation()

File: MatlabCodes/Chapter 05/Fig05_02.m
function MertonProcess_paths
clc;close all;

% Monte Carlo settings

NoOfPaths = 10;
NoOfSteps = 500;
T         = 5;

% Model parameters

r     = 0.05;
S0    = 100;
sigma = 0.2;
sigmaJ= 0.5;
muJ   = 0;
xiP   = 1;

[S,time,X] = GeneratePathsMerton(NoOfPaths,NoOfSteps,S0,T,xiP,muJ,sigmaJ,r,sigma);

figure(1)
plot(time,X,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$X(t)$$','interpreter','latex')
grid on

figure(2)
plot(time,S,'color',[0 0.45 0.74],'linewidth',1.2)
xlabel('time')
ylabel('$$S(t)$$','interpreter','latex')
grid on

% Martingale error

error = mean(S(:,end))-S0*exp(r*T)

function [S,time,X] =GeneratePathsMerton(NoOfPaths,NoOfSteps,S0,T,xiP,muJ,sigmaJ,r,sigma)

% Empty matrices for the Poisson process and stock paths

Xp     = zeros(NoOfPaths,NoOfSteps);
X      = zeros(NoOfPaths,NoOfSteps);
W      = zeros(NoOfPaths,NoOfSteps);
X(:,1) = log(S0);
S      = zeros(NoOfPaths,NoOfSteps);
S(:,1) = S0;
dt        = T/NoOfSteps;

% Random noise

Z1 = random('poisson',xiP*dt,[NoOfPaths,NoOfSteps]);
Z2 = random('normal',0,1,[NoOfPaths,NoOfSteps]);
J  = random('normal',muJ,sigmaJ,[NoOfPaths,NoOfSteps]);

% Creation of the paths

% Expectation E(exp(J))

EeJ = exp(muJ + 0.5*sigmaJ^2);
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps  
    if NoOfPaths>1
        Z2(:,i) = (Z2(:,i)-mean(Z2(:,i)))/std(Z2(:,i));
    end
    Xp(:,i+1) = Xp(:,i) + Z1(:,i);
    W(:,i+1)  = W(:,i)  + sqrt(dt)* Z2(:,i);
        
    X(:,i+1)  = X(:,i) + (r- xiP*(EeJ-1)-0.5*sigma^2)*dt +  sigma* (W(:,i+1)-W(:,i)) + J(:,i).*(Xp(:,i+1)-Xp(:,i));
    S(:,i+1) = exp(X(:,i));
    time(i+1) = time(i) + dt;
end
